File: Source/Gamemodes/Teams/TeamHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using WCSharp.Api;

public static class TeamHandler
{
    public static bool FreepickEnabled = false;

    public static void Handler(player player, int teamNumber, bool adminForced = false)
    {
        try
        {
            if (Gamemode.CurrentGameModeType == Globals.TEAM_MODES[0] && (adminForced || !RoundManager.GAME_STARTED && FreepickEnabled))
            {
                FreepickHandler(player, teamNumber);
            }
            else
            {
                player.DisplayTextTo($"{Colors.COLOR_YELLOW_ORANGE}The -team command is not available for this gamemode or the time to pick has expired.{Colors.COLOR_RESET}");
            }
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in TeamHandler.Handler: {e.Message}");
        }
    }

    private static void FreepickHandler(player player, int teamNumber)
    {
        if (CanPlayerJoinTeam(player, teamNumber))
        {
            ApplyPlayerToTeam(player, teamNumber);
        }
    }

    private static void ApplyPlayerToTeam(player player, int teamNumber)
    {
        if (TeamRegistry.TryGetTeam(teamNumber, out var team))
        {
            team.AddMember(player);
            player.DisplayTextTo($"{Colors.COLOR_YELLOW_ORANGE}You have joined team {team.TeamColor}{Colors.COLOR_RESET}");
        }
    }

    /// <summary>
    /// Throws all players who are currently not on a team into a random team. Prioritizing already made teams before creating new ones.
    /// </summary>
    public static void RandomHandler()
    {
        FreepickEnabled = false;
        var random = Globals.RANDOM_GEN;
        var shuffled = new List<player>(Globals.ALL_PLAYERS).OrderBy(x => random.Next()).ToList(); // seeded random shuffle, no desyncs -- this is only ever called once so its ok.
        var teamNumber = 1;

        try
        {
            foreach (var player in shuffled)
            {
                if (TeamRegistry.TryGetTeamForPlayer(player, out _)) continue;

                var addedToExistingTeam = false;
                foreach (var team in TeamRegistry.All)
                {
                    if (team.Teammembers.Count < Gamemode.PlayersPerTeam)
                    {
                        team.AddMember(player);
                        addedToExistingTeam = true;
                        break;
                    }
                }

                if (addedToExistingTeam) continue;

                // Create new teams as needed
                while (TeamRegistry.TryGetTeam(teamNumber, out var full) && full.Teammembers.Count >= Gamemode.PlayersPerTeam)
                {
                    teamNumber++;
                }

                if (!TeamRegistry.TryGetTeam(teamNumber, out var team2))
                {
                    team2 = new Team(teamNumber);
                }

                team2.AddMember(player);
            }
        }
        catch (Exception e)
        {
            Logger.Critical($"Error in TeamHandler.RandomHandler: {e.Message}");
        }
    }

    private static bool CanPlayerJoinTeam(player player, int teamNumber)
    {
        if (teamNumber > Globals.MAX_TEAM_SIZE)
        {
            player.DisplayTextTo($"{Colors.COLOR_YELLOW_ORANGE}Usage: -team 1-{Globals.MAX_TEAM_SIZE}{Colors.COLOR_RESET}");
            return false;
        }

        // If the team exists, check if it's full or if the player is already on it.
        if (TeamRegistry.TryGetTeam(teamNumber, out var team))
        {
            if (team.Teammembers.Count >= Gamemode.PlayersPerTeam)
            {
                player.DisplayTextTo($"{team.TeamColor}{Colors.COLOR_YELLOW_ORANGE} is full.{Colors.COLOR_RESET}");
                return false;
            }

            if (TeamRegistry.TryGetTeamForPlayer(player, out var currentTeam))
            {
                if (currentTeam.TeamID == teamNumber)
                {
                    player.DisplayTextTo($"{Colors.COLOR_YELLOW_ORANGE}You are already on {team.TeamColor}{Colors.COLOR_RESET}");
                    return false;
                }

                RemoveFromCurrentTeam(player);
            }
        }
        else
        {
            // Team doesn't exist yet — remove the player from their current team and create it.
            RemoveFromCurrentTeam(player);
            new Team(teamNumber);
        }
        return true;
    }

    private static void RemoveFromCurrentTeam(player player)
    {
        try
        {
            if (TeamRegistry.TryGetTeamForPlayer(player, out var team))
            {
                team.RemoveMember(player);
            }
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in TeamHandler.RemoveFromCurrentTeam: {e.Message}");
        }
    }
}

File: src/Gamemodes/Teams/TeamHandler.ts
import { Logger } from 'src/Events/Logger/Logger'
import { RoundManager } from 'src/Game/Rounds/RoundManager'
import { Globals } from 'src/Global/Globals'
import { GameSeed } from 'src/Init/GameSeed'
import { Colors } from 'src/Utility/Colors/Colors'
import { MapPlayer } from 'w3ts'
import { Gamemode } from '../Gamemode'
import { Team } from './Team'

export class TeamHandler {
    public static FreepickEnabled: boolean = false

    public static Handler(Player: MapPlayer, TeamNumber: number, adminForced: boolean = false) {
        if (
            Globals.CurrentGameModeType === Globals.TEAM_MODES[0] &&
            (adminForced || (!RoundManager.GAME_STARTED && this.FreepickEnabled))
        ) {
            this.FreepickHandler(Player, TeamNumber, adminForced)
        } else {
            Player.DisplayTextTo(
                `${Colors.COLOR_YELLOW_ORANGE}The -team command is not available for this gamemode or the time to pick has expired.${Colors.COLOR_RESET}`
            )
        }
    }

    private static FreepickHandler(Player: MapPlayer, TeamNumber: number, adminForced: boolean) {
        if (this.CanPlayerJoinTeam(Player, TeamNumber)) {
            this.ApplyPlayerToTeam(Player, TeamNumber)
        }
    }

    private static ApplyPlayerToTeam(Player: MapPlayer, TeamNumber: number) {
        const team = Globals.ALL_TEAMS.get(TeamNumber)
        if (team) {
            team.AddMember(Player)
            Player.DisplayTextTo(
                `${Colors.COLOR_YELLOW_ORANGE}You have joined team ${team.TeamColor}${Colors.COLOR_RESET}`
            )
        }
    }

    /// <summary>
    /// Throws all players who are currently not on a team into a random team. Prioritizing already made teams before creating new ones.
    /// </summary>
    public static RandomHandler() {
        this.FreepickEnabled = false
        let shuffled = GameSeed.Shuffle(Globals.ALL_PLAYERS, Globals.GAME_SEED)
        let teamNumber = 1
        try {
            for (let i: number = 0; i < shuffled.length; i++) {
                let player = shuffled[i]

                let currentTeam = Globals.PLAYERS_TEAMS.get(player)
                if (currentTeam) {
                    continue
                }
                let addedToExistingTeam: boolean = false

                // Attempt to add player to an existing team
                for (let j: number = 0; j < Globals.ALL_TEAMS_LIST.length; j++) {
                    let team = Globals.ALL_TEAMS_LIST[j]
                    if (team.Teammembers.length < Gamemode.PlayersPerTeam) {
                        team.AddMember(player)
                        addedToExistingTeam = true
                        break
                    }
                }

                if (!addedToExistingTeam) {
                    // Create new teams as needed
                    while (
                        Globals.ALL_TEAMS.has(teamNumber) &&
                        Globals.ALL_TEAMS.get(teamNumber)!.Teammembers.length >= Gamemode.PlayersPerTeam
                    ) {
                        teamNumber++
                    }

                    // Check if the team exists, if not create it
                    let team = Globals.ALL_TEAMS.get(teamNumber)
                    if (!team) {
                        team = new Team(teamNumber)
                    }

                    // Add the player to the new team
                    team.AddMember(player)
                }
            }
        } catch (e: any) {
            Logger.Critical(`Error in TeamHandler.RandomHandler: ${e}`)
        }
    }

    private static CanPlayerJoinTeam(Player: MapPlayer, TeamNumber: number) {
        if (TeamNumber > 24) {
            Player.DisplayTextTo(`${Colors.COLOR_YELLOW_ORANGE}Usage: -team 1-24${Colors.COLOR_RESET}`)
            return false
        }

        // If the team exists, we're going to check if full or if that player is already on the team
        let team = Globals.ALL_TEAMS.get(TeamNumber)
        if (team) {
            // If Team is full, return.
            if (team.Teammembers.length >= Gamemode.PlayersPerTeam) {
                Player.DisplayTextTo(`${team.TeamColor}${Colors.COLOR_YELLOW_ORANGE} is full.${Colors.COLOR_RESET}`)
                return false
            }
            // If player is on the team they're trying to join.. Return.
            let currentTeam = Globals.PLAYERS_TEAMS.get(Player)
            if (currentTeam) {
                if (currentTeam.TeamID === TeamNumber) {
                    Player.DisplayTextTo(
                        `${Colors.COLOR_YELLOW_ORANGE}You are already on ${team.TeamColor}${Colors.COLOR_RESET}`
                    )
                    return false
                }

                // If not the same team.. Remove them so they're ready to join another.
                this.RemoveFromCurrentTeam(Player)
            }
        }
        // If team doesnt exist, we're going to remove the player from current team and create that new team.
        else {
            this.RemoveFromCurrentTeam(Player)
            new Team(TeamNumber)
        }
        return true
    }

    private static RemoveFromCurrentTeam(Player: MapPlayer) {
        let team = Globals.PLAYERS_TEAMS.get(Player)
        if (team) {
            team.RemoveMember(Player)
        }
    }
}
